drivers/adc/adc_npcx.c

```diff
@@ -0,0 +1,376 @@
+	const struct device *adc_dev;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) \
+	((const struct adc_npcx_config *)(dev)->config)
+
+#define DRV_DATA(dev) \
+	((struct adc_npcx_data *)(dev)->data)
+
+#define HAL_INSTANCE(dev) \
+	(struct adc_reg *)(DRV_CONFIG(dev)->base)
+
+/* ADC local functions */
+static void adc_npcx_isr(void *arg)
+{
+	struct adc_npcx_data *const data = DRV_DATA((const struct device *)arg);
+	struct adc_reg *const inst = HAL_INSTANCE((const struct device *)arg);
+	uint16_t status = inst->ADCSTS;
+	uint16_t result, channel;
+
+	/* Clear status pending bits first */
+	inst->ADCSTS = status;
+	LOG_DBG("%s: status is %04X\n", __func__, status);
+
+	/* Is end of conversion cycke event? ie. Scan conversion is done. */
+	if (IS_BIT_SET(status, NPCX_ADCSTS_EOCCEV)) {
+		/* Stop conversion for scan conversion mode */
+		inst->ADCCNF |= BIT(NPCX_ADCCNF_STOP);
+
+		/* Get result for each ADC selected channel */
+		while (data->channels) {
+			channel = find_lsb_set(data->channels) - 1;
+			result = GET_FIELD(inst->CHNDAT[channel],
+					NPCX_CHNDAT_CHDAT_FIELD);
+			/* Save ADC result */
+			*data->buffer++ = result;
+			data->channels &= ~BIT(channel);
+		}
+
+		/* Turn off ADC and inform sampling is done */
+		inst->ADCCNF &= ~(BIT(NPCX_ADCCNF_ADCEN));
+		adc_context_on_sampling_done(&data->ctx, data->adc_dev);
+	}
+}
+
+static int adc_npcx_validate_buffer_size(const struct device *dev,
+					const struct adc_sequence *sequence)
+{
+	uint8_t channels = 0;
+	uint32_t mask;
+	size_t needed;
+
+	for (mask = BIT(NPCX_ADC_CH_COUNT - 1); mask != 0; mask >>= 1) {
+		if (mask & sequence->channels) {
+			channels++;
+		}
+	}
+
+	needed = channels * sizeof(uint16_t);
+	if (sequence->options) {
+		needed *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static void adc_npcx_start_scan(const struct device *dev)
+{
+	struct adc_npcx_data *const data = DRV_DATA(dev);
+	struct adc_reg *const inst = HAL_INSTANCE(dev);
+
+	/* Turn on ADC first */
+	inst->ADCCNF |= BIT(NPCX_ADCCNF_ADCEN);
+
+	/* Update selected channels in scan mode by channels mask */
+	inst->ADCCS = data->channels;
+
+	/* Select 'Scan' Conversion mode. */
+	SET_FIELD(inst->ADCCNF, NPCX_ADCCNF_ADCMD_FIELD,
+			NPCX_ADC_SCAN_CONVERSION_MODE);
+
+	/* Select 'One-Shot' Repetitive mode */
+	inst->ADCCNF |= BIT(NPCX_ADCCNF_INTECEN);
+
+	/* Start conversion */
+	inst->ADCCNF |= BIT(NPCX_ADCCNF_START);
+
+	LOG_DBG("Start ADC scan conversion and ADCCNF,ADCCS are (%04X,%04X)\n",
+			inst->ADCCNF, inst->ADCCS);
+}
+
+static int adc_npcx_start_read(const struct device *dev,
+					const struct adc_sequence *sequence)
+{
+	struct adc_npcx_data *const data = DRV_DATA(dev);
+	int error = 0;
+
+	if (!sequence->channels ||
+	    (sequence->channels & ~BIT_MASK(NPCX_ADC_CH_COUNT))) {
+		LOG_ERR("Invalid ADC of channels");
```

Comment:
This message seems to need a correction.

Reply:
Done.